src/main/lib/osu-file-parser/OsuParser.ts

```diff
@@ -1,125 +1,345 @@
+        song.title = db.read_string().trim();
+        song.titleUnicode = db.read_string().trim();
+        song.creator = db.read_string().trim();
+
+        // NOTE: I'm being lazy here, and only loading the first diff for a given audio file.
+        //       This is wrong, especially if diffs of a set aren't right next to each other in the db.
+        const diff_name = db.read_string();
+        song.diffs = [diff_name];
+
+        const audio_filename = db.read_string();
+
+        db.read_hash(); // .osu file md5
+        const osu_filename = db.read_string().trim();
+        db.read_u8(); // ranking status
+        db.read_u16(); // nb circles
+        db.read_u16(); // nb sliders
+        db.read_u16(); // nb spinners
+
+        // TODO: This probably is incorrect. See alternatives:
+        // tms_a = (tms - 621355968000000000) / 10000000
+        // tms_b = (tms - 504911232000000000)
+        const last_modification_time = db.read_u64();
+        song.dateAdded = new Date(last_modification_time).toISOString();
+
+        db.read_f32(); // AR
+        db.read_f32(); // CS
+        db.read_f32(); // HP
+        db.read_f32(); // OD
+        db.read_f64(); // slider multiplier
+
+        // std
+        let nb_star_ratings = db.read_u32();
+        for (let s = 0; s < nb_star_ratings; s++) {
+          db.read_u8();
+          db.read_u32(); // mod flags
+          db.read_u8();
+          db.read_f64(); // star rating
+        }
 
-          if (diff.isError) {
-            continue;
-          }
+        // taiko
+        nb_star_ratings = db.read_u32();
+        for (let s = 0; s < nb_star_ratings; s++) {
+          db.read_u8();
+          db.read_u32(); // mod flags
+          db.read_u8();
+          db.read_f64(); // star rating
+        }
 
-          s.diffs.push(diff.value.toLowerCase());
-          continue;
+        // ctb
+        nb_star_ratings = db.read_u32();
+        for (let s = 0; s < nb_star_ratings; s++) {
+          db.read_u8();
+          db.read_u32(); // mod flags
+          db.read_u8();
+          db.read_f64(); // star rating
         }
 
-        const parsed = await OsuParser.parseFile(file);
-        if (parsed.isError) {
-          continue;
+        // mania
+        nb_star_ratings = db.read_u32();
+        for (let s = 0; s < nb_star_ratings; s++) {
+          db.read_u8();
+          db.read_u32(); // mod flags
+          db.read_u8();
+          db.read_f64(); // star rating
         }
 
-        const result = await parsed.value.toSong(dir);
-        if (result.isError) {
-          continue;
+        db.read_u32(); // drain time
+        song.duration = db.read_u32() / 1000.0;
+        db.read_u32(); // preview time
+
+        const nb_timing_points = db.read_u32();
+        song.bpm = [];
+        for (let t = 0; t < nb_timing_points; t++) {
+          const ms_per_beat = db.read_f64();
+          const offset = db.read_f64();
+          const timing_change = !!db.read_u8();
```

Comment:
Remove if unused. Also what is this boolean conversion `!!db.read_u8();`, `Boolean(db.read_u8())` can be used instead, right?

Reply:
removed